### ISSUE TITLE
Query not working as expected, returns repeat labels with difference distances.

### ISSUE THREAD

**Issue by ShuAiii** [nmslib/hnswlib#313]
I ran the method `add_items`, `knn_query` then `add_items` for a second time. On the second time of attempting to run `knn_query`, it looks like the graph was corrupted.

```
>>> knn.space
'l2'
>>> knn.dim
391
>>> knn.M
48
>>> knn.ef_construction
50
>>> knn.max_elements
300000000
>>> knn.element_count
118698261
>>> knn.knn_query(point, k = 2, num_threads = 1)
(array([[    0, 22435]], dtype=uint64), array([[34.451694, 61.909622]], dtype=float32))
>>> knn.knn_query(point, k = 5, num_threads = 1)
(array([[    0, 22435,     0,     0,     0]], dtype=uint64), array([[ 34.451694,  61.909622, 134.31993 , 134.31993 , 134.31993 ]],
      dtype=float32))
>>> knn.knn_query(point, k = 10, num_threads = 1)
(array([[    0, 22435,     0,     0,     0,     0,     0,     0,     0,
            0]], dtype=uint64), array([[ 34.451694,  61.909622, 134.31993 , 134.31993 , 134.31993 ,
        134.31993 , 134.31993 , 134.31993 , 134.31993 , 134.31993 ]],
      dtype=float32))
>>> knn.set_ef(40)
>>> knn.knn_query(point, k = 10, num_threads = 1)
(array(**[[    0, 22435,     0,     0,     0,     0,     0,     0,     0,
            0]], dtype=uint64), array([[ 34.451694,  61.909622, 134.31993 , 134.31993 , 134.31993 ,
        134.31993 , 134.31993 , 134.31993 , 134.31993 , 134.31993 ]]**,
      dtype=float32))
>>> 
```

As you see here, the label 0 appears repetitively as NN, and the first and third instance yields different distance values.

**Comment by yurymalkov** [comment]
Hm. Do you add new element or update existing ones? Also, do you have duplicates in the dataset?

**Comment by ShuAiii** [comment]
@yurymalkov Yeah I did. It turns out, out of the 118698261 elements I added, only about 120k are none zeros. Rest are all zero vectors.

I did use `pickle.dumps()` and `pickle.load()` to save and reload Index many times, then add more points after reload. Could this be related to the not thread-safe issue? Has anyone else observed this zero-vectors symptom too?

**Comment by yurymalkov** [comment]
Yes, this can easily be thread-safe issues. No, I do not think there it was noticed before.
Do you add/save items in parallel?

**Comment by ShuAiii** [comment]
I did not. I ran `add_items` and `pickle.dumps()` separately, always.

On the bright-side, I start a new experiment and swapped out  `pickle` for your "thread-safe" method `save_index()` and `load_index()`. The zero vector problem disappeared, however I'm seeing this:
`
>>> knn = hnswlib.Index("l2", 10)
>>> knn.init_index(max_elements=1000, ef_construction = 200, M = 16, random_seed = 100)
>>> knn.add_items(data, num_threads=1)
>>> knn.save_index("knn.bin")
>>> new_knn = hnswlib.Index(space='l2', dim=10)
>>> new_knn.load_index("knn.bin", max_elements = 1000)
>>> knn.get_items(knn.get_ids_list())
[[1.0, 23.0, 4.0, 5.0, 6.0, 432.0, 3.0, 4.0, 5.0, 6.0]]
>>> new_knn.get_items(new_knn.get_ids_list())
[[1.0, 23.0, 4.0, 5.0, 6.0, 432.0, 3.0, 4.0, 5.0, 6.0]]
>>> knn.M
16
>>> new_knn.M
0
>>> knn.ef_construction
200
>>> new_knn.ef_construction
0
>>> knn.space
'l2'
>>> new_knn.space
'l2'
>>> knn.dim
10
>>> new_knn.dim
10
>>> knn.element_count
1
>>> new_knn.element_count
0
>>> knn.ef
10
>>> new_knn.ef
10
`

**Comment by ShuAiii** [comment]
I'll mark this issue as closed, since the alternative index saving method to pickle seems to do it's trick.

I'll open the setter as a new issue.
